staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go

```diff
@@ -349,14 +359,85 @@ func validateClaimMappings(compiler authenticationcel.Compiler, celMapper *authe
 		}
 	}
 
+	if structuredAuthnFeatureEnabled && needsToVerifyEmail && !hasVerifiedEmail && !anyUsesEmailVerifiedClaim(extraCompilationResults) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("username", "expression"), m.Username.Expression,
+			"claims.email_verified must be used in claimMappings.extra[*].valueExpression or claimValidationRules[*].expression when claims.email is used in the username expression"))
+	}
+
 	if structuredAuthnFeatureEnabled && len(extraCompilationResults) > 0 {
 		celMapper.Extra = authenticationcel.NewClaimsMapper(extraCompilationResults)
 	}
 
 	return allErrs
 }
 
-func validatePrefixClaimOrExpression(compiler authenticationcel.Compiler, mapping api.PrefixedClaimOrExpression, fldPath *field.Path, claimOrExpressionRequired, structuredAuthnFeatureEnabled bool) (*authenticationcel.CompilationResult, field.ErrorList) {
+func usesEmailClaim(ast *celgo.Ast) bool {
+	return hasSelectExp(ast.Expr(), "claims", "email")
+}
+
+func anyUsesEmailVerifiedClaim(results []authenticationcel.CompilationResult) bool {
+	for _, result := range results {
+		if usesEmailVerifiedClaim(result.AST) {
+			return true
+		}
+	}
+	return false
+}
+
+func usesEmailVerifiedClaim(ast *celgo.Ast) bool {
+	return hasSelectExp(ast.Expr(), "claims", "email_verified")
+}
+
+func hasSelectExp(exp *exprpb.Expr, operand, field string) bool {
```

Comment:
For reference, here's how we statically checked if a CEL expression accesses the `oldSelf` variable for transition rules: https://github.com/kubernetes/kubernetes/blob/a5f5f44157c49fdfb6384862c7cb34c2ddbd4cce/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go#L218-L229

This case is different because it's not just checking if a root variable is referenced, so I don't know if it can work here. But I just wanted to check if this approach has been explored and rejected?

Reply:
@jpbetz that only shows `claims` as being used:

<img width="450" alt="image" src="https://github.com/kubernetes/kubernetes/assets/7001223/9c19527f-cef9-4f44-bcd0-e839853a2842">

Reply:
Gotcha. Can we leverage [`Visitor()`](https://github.com/google/cel-go/blob/9775e65f751b9d2e04d8f942f832de889ec3fb14/common/ast/navigable.go#L139) to avoid reimplementing [visit](https://github.com/google/cel-go/blob/9775e65f751b9d2e04d8f942f832de889ec3fb14/common/ast/navigable.go#L194)?

Reply:
^ @TristonianJones for any CEL internals insights

Reply:
@jpbetz so I gave that a shot and encountered a few issues:

- Requires cel-go to be bumped to v0.18.0 since that is the earliest tag with the visitor code
- The bump adds/changes a lot of code, and a few places in the code stop compiling in ways that is not super obvious how to fix (I hacked around this)
- The AST type we have contains a private AST implementation type, which is actually what we need to use the visitor.  There doesn't appear to be any clean way for me to get to the underlying type (I also hacked around this)

The visitor code does appear to work, but the other issues make it a rather unsafe choice this late in the release cycle.

Reply:
Thanks for trying. We'll circle back on this one.

Reply:
cc @cici37

Reply:
Hi @enj, we shifted over to using the `ast` package so you can avoid the direct proto dep:

```go
root := ast.NavigateAST(resultAST.NativeRep())
matches := ast.MatchDescendents(root, <matcher>)
if len(matches) > 0 { ... }
``` 

The `cel/folding.go` implementation is a pretty good reference point for navigating expressions.

Reply:
It looks like we would need `v0.18.2` to use the `NativeRep` method.  Either way I think we will have to wait to bump the dep until next release.  I will take a look into `cel/folding.go` and make sure my approach lines up with it.